guidance/models/_anthropic.py

```diff
@@ -52,24 +45,73 @@ def __init__(
 
     def _generator(self, prompt, temperature):
 
+        # find the role tags
+        pos = 0
+        role_end = b"<|im_end|>\n"
+        messages = []
+        found = True
+        system_prompt = None # Not mandatory, but we'll store it if found
+        while found:
+
+            # find the role text blocks
+            found = False
+            for role_name, start_bytes in (
+                ("system", b"<|im_start|>system\n"),
+                ("user", b"<|im_start|>user\n"),
+                ("assistant", b"<|im_start|>assistant\n"),
+            ):
+                if prompt[pos:].startswith(start_bytes):
+                    pos += len(start_bytes)
+                    end_pos = prompt[pos:].find(role_end)
+                    if end_pos < 0:
+                        assert (
+                            role_name == "assistant"
+                        ), "Bad chat format! Last role before gen needs to be assistant!"
+                        break
+                    btext = prompt[pos : pos + end_pos]
+                    pos += end_pos + len(role_end)
+                    if role_name == "system":
+                        system_prompt = btext.decode("utf8")
+                    else:
+                        messages.append(
+                            {"role": role_name, "content": btext.decode("utf8")}
+                        )
+                    found = True
+                    break
+
+        # Add nice exception if no role tags were used in the prompt.
+        # TODO: Move this somewhere more general for all chat models?
+        if messages == []:
+            raise ValueError(
+                f"The AnthropicAI model {self.model_name} is a Chat-based model and requires role tags in the prompt! \
+            Make sure you are using guidance context managers like `with system():`, `with user():` and `with assistant():` \
+            to appropriately format your guidance program for this type of model."
+            )
+
         # update our shared data state
         self._reset_shared_data(prompt, temperature)
 
+        # API call and response handling
         try:
-            generator = self.anthropic.completions.create(
+            # Need to do this because Anthropic API is a bit weird with the system keyword...
+            model_kwargs = dict(
```

Comment:
Not absolutely necessary, but can you get token stats back from the models as well?

Reply:
I don't think the Anthropic streaming api supports token usage stats either :( https://docs.anthropic.com/en/api/messages-streaming

Only OpenAI seems to for now.

It might be that each chunk we get sent back is exactly one token? But anthropic's tokenizer is not public so it's also hard to verify this. And there's no way to count the prompt tokens as far as I can tell (without making another API call to a non-streaming version of the API)

Reply:
If they don't support it, they don't support it. If we really wanted approximate metrics, we could grab another tokeniser, but we'd have to warn users that things were approximate. But that's not a discussion for this PR.